packages/SwingSet/src/vats/timer/vat-timer.js

```diff
@@ -1,98 +1,977 @@
+        return;
+      }
+      // repeaters stay in "waiting" until their promise resolves,
+      // at which point we either reschedule or cancel
+      E(handler)
+        .wake(getCurrentTimestamp())
+        .then(
+          _res => self.rescheduleYourself(),
+          _err => self.cancel(),
+        )
+        .catch(err => console.log(`timer repeater error`, err));
+    },
+
+    cancel({ self, state }) {
+      const { scheduled, cancelToken } = state;
+      removeCancel(cancels, cancelToken, self);
+      state.cancelled = true;
+      if (scheduled) {
+        removeEvent(schedule, scheduled, self);
+        state.scheduled = undefined;
+        reschedule();
+      }
+    },
+  };
+
+  const makeRepeaterEvent = vivifyKind(
+    baggage,
+    'repeaterEvent',
+    initRepeaterEvent,
+    repeaterEventBehavior,
+  );
+
+  // -- more helper functions
+
+  /**
+   * @param {TimestampValue} when
+   * @param {CancelToken} cancelToken
+   * @returns {Promise<Timestamp>}
+   */
+  function wakeAtInternal(when, cancelToken) {
+    const event = makePromiseEvent(when, cancelToken);
+    const { resolve, reject, promise } = makePromiseKit();
+    // these 'controls' are never shared off-vat, but we wrap them as
+    // Far to appease WeakMapStore's value requirements
+    const controls = Far('controls', { resolve, reject });
+    wakeupPromises.init(event, controls);
+    event.scheduleYourself();
+    return promise; // disconnects upon upgrade
+  }
+
+  // -- now we can define the public API functions
+
+  /**
+   * @returns {Timestamp}
+   */
+  function getCurrentTimestamp() {
+    return toTimestamp(getNow());
+  }
+
+  /**
+   * @param {Timestamp} when
+   * @param {Handler} handler
+   * @param {CancelToken} [cancelToken]
+   * @returns {Timestamp}
+   */
+  function setWakeup(when, handler, cancelToken = undefined) {
+    when = fromTimestamp(when);
+    assert.equal(passStyleOf(handler), 'remotable', 'bad setWakeup() handler');
+    if (cancelToken) {
+      assert.equal(passStyleOf(cancelToken), 'remotable', 'bad cancel token');
+    }
+
+    const event = makeOneShotEvent(when, handler, cancelToken);
+    event.scheduleYourself();
+    // TODO this is the documented behavior, but is it useful?
+    return toTimestamp(when);
+  }
+
+  /**
+   * wakeAt(when): return a Promise that fires (with the scheduled
+   * wakeup time) somewhat after 'when'. If a 'cancelToken' is
+   * provided, calling ts.cancel(cancelToken) before wakeup will cause
+   * the Promise to be rejected instead.
+   *
+   * @param {Timestamp} when
+   * @param {CancelToken} [cancelToken]
+   * @returns { Promise<Timestamp> }
+   */
+  function wakeAt(when, cancelToken = undefined) {
+    when = fromTimestamp(when);
+    const now = getNow();
+    if (when <= now) {
+      return Promise.resolve(toTimestamp(now));
+    }
+    return wakeAtInternal(when, cancelToken);
+  }
+
+  /**
+   * delay(delay): return a Promise that fires (with the scheduled
```

Comment:
Oh, hm, so here the function is defined as `addDelay`, but when exported on `TimerService` it's named `delay`. I changed the local function name to avoid confusion with the `delay` parameter that it takes, and so the reader can correlate it with the external API. Which way do you think I should spell it in the comment? And, would it be a good idea to move this to `const delay = (delay, cancelToken) => ..` instead of `const addDelay = ..` ? The visual confusion would still exist, and I think eslint would still complain about the scope shadowing...

Reply:
The comment should reflect the local name.  If you want, it would be fine to add a parenthetical remark that it will be exported as `delay(delay, cancelToken)`.

I don't think `const delay = (delay, ...)` would be better.